### DISCUSSION TITLE
What about helper contracts in foundry

### DISCUSSION THREAD

**Discussion by khushbusandhu** [Cyfrin/foundry-full-course-cu#752] [General]
Why does Helper contracts are used in foundry?

**Comment by theirrationalone** [comment]
If you mean `LinkToken` Contract or something like it...

We wanna interact with some off-chain protocols to get some real world data or/and some logic that facilitates to achieve our goal. So to communicate/interact with them we can create a contract, interface, library, etc, Which in turn generally makes an `api call` by pointing intended protocol/contract to get intended data as well as logic.

**Reply by khushbusandhu** [reply]
no, I am asking about the helperConfig contracts as I am not getting the point of creating that contract when we have our main contract.
and according to chat GPT creating these contracts are used to modularize and optimize code. but how? How can an external contract increase the code reusability and make it easy to understand

**Reply by theirrationalone** [reply]
Ohhhh, You mean outsourcing....

It's a very popular decoration in programming world to outsource repeatedly occurring task. Let's take an example of our foundry-raffle/lottery....

```solidity
struct NetworkConfig {
        address vrfCoordinatorV2Address;
        uint256 entranceFee;
        bytes32 gasLane;
        uint64 subId;
        uint32 callbackGasLimit;
        uint256 interval;
        uint256 privateKey;
        address linkTokenAddress;
    }
```

Now Go to Lottery Project and inspect how often, how many times you needed these data together or one at an instance and count how many times in how many places you initialized a new `HelperConfig`contract. 

So It's in front at our eyes, How needy these data are i.e., vrfcoordinator, gasLane, subId, etc,....

Let's say for a bit ... just assume we don't have `HelperConfig`....
Since we don't have `HelperConfig`, Now we have to repeat the task given below whenever we needed that data...

```solidity
        vm.startBroadcast(ANVIL_PRIVATE_KEY);
        VRFCoordinatorV2Mock vrfCoordinatorV2Mock = new VRFCoordinatorV2Mock(BASE_FEE, GAS_PRICE_LINK);
        LinkToken linkToken = new LinkToken();
        vm.stopBroadcast();

        return NetworkConfig({
            vrfCoordinatorV2Address: address(vrfCoordinatorV2Mock),
            entranceFee: ENTRANCE_FEE,
            gasLane: 0x8af398995b04c28e9951adb9721ef74c74f93e6a478f39e7e0777be13527e7ef,
            subId: 0,
            callbackGasLimit: 5000000,
            interval: 30,
            privateKey: ANVIL_PRIVATE_KEY,
            linkTokenAddress: address(linkToken)
        });
```

So It's considered as The Best Practice to `Write Once, Use X Times`
As a side effect, Might Curiosity started igniting into your head 😅, Go again and start hunting replicated - repeated big code statements and outsource/modularize em'all.

See I asked the same question from `ChatGpt`:

![Screenshot (133)](https://github.com/Cyfrin/foundry-full-course-f23/assets/96755510/85d0f944-28d8-404a-80b3-2e57c93d29e7)

**Reply by khushbusandhu** [reply]
okay I got it
thanks @theirrationalone
